### PR TITLE
Multicluster Adding Delete logic for dynamically created controllers

### PR DESCRIPTION
**Pull request by sbezverk**
This PR adds delete logic to secret controller. On detection of remote cluster's secret removal, 1: corresponding service registry entry gets deleted, 2: Stop message sent to the controller, 3: remote cluster gets removed from the clusters store.
Closes: #5629
Closes: #5622
Closes: #5606
Closes: #5605

### CODE REVIEW

**Comment by sbezverk** [comment]
@gyliu513 I added delete logic for dynamically create controllers, I have tested it from control plane perspective, I would appreciate if you could test it from data plane as well.

**Comment by sbezverk** [comment]
@baodongli I have run several tests and I do not hit any issue on a dynamic controller deletion. Appreciate a lot your comments.

File: pilot/pkg/serviceregistry/aggregate/controller.go [hunk 1]
```diff
@@ -45,28 +45,67 @@ type Controller struct {
 
 // NewController creates a new Aggregate controller
 func NewController() *Controller {
+	log.Infof("><SB> Aggregate Controller NewController was called")
```
Comment (costinm) [inline pilot/pkg/serviceregistry/aggregate/controller.go]: Remove

File: pilot/pkg/serviceregistry/aggregate/controller.go [hunk 2]
```diff
@@ -45,28 +45,67 @@ type Controller struct {
...
+	defer c.storeLock.Unlock()
+
+	registries := make([]Registry, len(c.registries))
+	copy(registries, c.registries)
+	return registries
+}
+
+// GetRegistryIndex returns the index of a registry
+func (c *Controller) GetRegistryIndex(clusterID string) (int, bool) {
+	for i, r := range c.registries {
```
Comment (costinm) [inline pilot/pkg/serviceregistry/aggregate/controller.go]: Still not happy. Can you instead copy the registries when you add/remove - and return the same slice, without copy, in get ? Copy is rare event, GetRegistry is extremely frequent. Memory allocations/GC do matter.

Comment (costinm) [reply pilot/pkg/serviceregistry/aggregate/controller.go]: I mean add/remove is rare.

File: pilot/pkg/config/clusterregistry/conversion.go [hunk 3]
```diff
@@ -42,33 +43,50 @@ const (
...
+}
+
 // ClusterStore is a collection of clusters
 type ClusterStore struct {
-	clusters      []*k8s_cr.Cluster
-	clientConfigs map[string]clientcmdapi.Config
-	storeLock     sync.RWMutex
+	rc map[Metadata]*RemoteCluster
+
+	//	clusters      []*k8s_cr.Cluster
```
Comment (gyliu513) [inline pilot/pkg/config/clusterregistry/conversion.go]: we can remove those two useless parameters

File: pilot/pkg/config/clusterregistry/conversion.go [hunk 4]
```diff
@@ -42,33 +43,50 @@ const (
...
-		clientConfigs: map[string]clientcmdapi.Config{},
+		rc: rc,
 	}
 }
 
 // GetClientAccessConfigs returns map of collected client configs
-func (cs *ClusterStore) GetClientAccessConfigs() map[string]clientcmdapi.Config {
-	return cs.clientConfigs
-}
+//func (cs *ClusterStore) GetClientAccessConfigs() map[string]clientcmdapi.Config {
```
Comment (gyliu513) [inline pilot/pkg/config/clusterregistry/conversion.go]: remove this as well

File: pilot/pkg/serviceregistry/aggregate/controller.go [hunk 5]
```diff
@@ -45,16 +45,57 @@ type Controller struct {
...
+
+	if len(c.registries) == 0 {
+		log.Warnf("Registry list is empty, nothing to delete")
+		return
+	}
+	index, ok := c.GetRegistryIndex(clusterID)
+	if !ok {
+		log.Warnf("Registry is not found in the registries list, nothing to delete")
+		return
+	}
```
Comment (gyliu513) [inline pilot/pkg/serviceregistry/aggregate/controller.go]: I'd like a log here to clarify which cluster is being removed from registry

File: pilot/pkg/config/clusterregistry/conversion.go (hunk 4 above)
Comment (sbezverk) [reply pilot/pkg/config/clusterregistry/conversion.go]: done

File: pilot/pkg/serviceregistry/aggregate/controller.go (hunk 5 above)
Comment (sbezverk) [reply pilot/pkg/serviceregistry/aggregate/controller.go]: done

**Comment by sbezverk** [comment]
@gyliu513 thanks a lot for your review and tests.

File: pilot/pkg/config/clusterregistry/secretcontroller.go [hunk 6]
```diff
@@ -244,31 +246,32 @@ func addMemberCluster(s *corev1.Secret, c *Controller) {
...
-			if c.ObjectMeta.Name == s {
-				cs.clusters = append(cs.clusters[:i], cs.clusters[i+1:]...)
-				break
-			}
-		}
+		// Deleting Service registry associated with controller
+		c.serviceController.DeleteRegistry(GetClusterID(c.cs.rc[s].Cluster))
+		// Stop controller
+		close(c.cs.rc[s].ControlChannel)
+		<-c.cs.rc[s].ControlChannel
```
Comment (baodongli) [inline pilot/pkg/config/clusterregistry/secretcontroller.go]: can't receive after the channel has been closed.

File: pilot/pkg/config/clusterregistry/secretcontroller.go [hunk 7]
```diff
@@ -244,31 +246,32 @@ func addMemberCluster(s *corev1.Secret, c *Controller) {
 				Controller:       kubectl,
 			})
 		stopCh := make(chan struct{})
+		c.cs.rc[key].ControlChannel = stopCh
```
Comment (baodongli) [inline pilot/pkg/config/clusterregistry/secretcontroller.go]: If you look at the stop channel for the existing controllers, the go routine associated with the controller terminates when the stop channel is closed. See the WaitSignal() call in pilot-discovery/main.go. I think that that logic should also apply to the controllers dynamically created for multi-cluster (in addition to this additional stopch created at line 249).

File: pilot/pkg/serviceregistry/aggregate/controller.go [hunk 8]
```diff
@@ -45,16 +45,58 @@ type Controller struct {
...
+	c.registries = registries
+	log.Infof("Registry for the cluster %s has been deleted.", clusterID)
+}
+
+// GetRegistries returns a copy of all registries
+func (c *Controller) GetRegistries() []Registry {
+	c.storeLock.Lock()
+	defer c.storeLock.Unlock()
+
+	return c.registries
```
Comment (baodongli) [inline pilot/pkg/serviceregistry/aggregate/controller.go]: I still think that the lock should also protect the access to individual controller. The question is whether or not the controller would be stopped by DeleteRegistry() while it's being queried. If the answer is yes, then GetRegistries() can't be used as is.

File: pilot/pkg/config/clusterregistry/secretcontroller.go (hunk 6 above)
Comment (sbezverk) [reply pilot/pkg/config/clusterregistry/secretcontroller.go]: This is the normal practice of informing go routine to stop via stop channel and waiting until that go routine exits.

File: pilot/pkg/config/clusterregistry/secretcontroller.go (hunk 7 above)
Comment (sbezverk) [reply pilot/pkg/config/clusterregistry/secretcontroller.go]: As I mentioned above, this is a standard way of informing go routine to stop via stop channel. Please try it and you will see it works as expected.

**Comment by sbezverk** [comment]
@costin I do not know why but you merged only 1 commit from this PR, there were several additional changes missing in the merge. Please either revert previous merged incomplete change, or merge this complete change..

**Pull request merged**